flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateUtil.scala

```diff
@@ -0,0 +1,309 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.api.table.runtime.aggregate
+
+import java.util
+
+import org.apache.calcite.rel.`type`._
+import org.apache.calcite.rel.core.AggregateCall
+import org.apache.calcite.sql.SqlAggFunction
+import org.apache.calcite.sql.`type`.SqlTypeName._
+import org.apache.calcite.sql.`type`.{SqlTypeFactoryImpl, SqlTypeName}
+import org.apache.calcite.sql.fun._
+import org.apache.flink.api.common.functions.{GroupReduceFunction, MapFunction}
+import org.apache.flink.api.table.Row
+import org.apache.flink.api.table.plan.PlanGenException
+import org.apache.flink.api.table.plan.nodes.logical.FlinkAggregate
+
+import scala.collection.JavaConversions._
+import scala.collection.mutable.ArrayBuffer
+
+object AggregateUtil {
+
+  /**
+   * Create Flink operator functions for aggregates. It includes 2 implementations of Flink 
+   * operator functions:
+   * [[org.apache.flink.api.common.functions.MapFunction]] and 
+   * [[org.apache.flink.api.common.functions.GroupReduceFunction]](if it's partial aggregate,
+   * should also implement [[org.apache.flink.api.common.functions.CombineFunction]] as well). 
+   * The output of [[org.apache.flink.api.common.functions.MapFunction]] contains the 
+   * intermediate aggregate values of all aggregate function, it's stored in Row by the following
+   * format:
+   *
+   * {{{
+   *                   avg(x) aggOffsetInRow = 2          count(z) aggOffsetInRow = 5
+   *                             |                          |
+   *                             v                          v
+   *        +---------+---------+--------+--------+--------+--------+
+   *        |groupKey1|groupKey2|  sum1  | count1 |  sum2  | count2 |
+   *        +---------+---------+--------+--------+--------+--------+
+   *                                              ^
+   *                                              |
+   *                               sum(y) aggOffsetInRow = 4
+   * }}}
+   *
+   */
+  def createOperatorFunctionsForAggregates(aggregate: FlinkAggregate,
+      inputType: RelDataType, outputType: RelDataType,
+      groupings: Array[Int]): AggregateResult = {
+
+    val aggregateCalls: Seq[AggregateCall] = aggregate.getAggCallList
+    // store the aggregate fields of each aggregate function, by the same order of aggregates.
+    val aggFieldIndexes = new Array[Int](aggregateCalls.size)
+    val aggregates = new Array[Aggregate[_ <: Any]](aggregateCalls.size)
+
+    transformToAggregateFunctions(aggregateCalls, aggFieldIndexes,
```

Comment:
It's better Scala style to define a function that returns values. If, as in this case, two values are returned, you can return a tuple.
```
private def transformToAggregateFunctions(
  aggregateCalls: Seq[AggregateCall], 
  inputType: RelDataType, 
  groupKeysCount: Int): (Array[Int], Array[Aggregate[_ <: Any]]) = 
{
  // ...
}
```